File: December 2024 Leetcode Solution/December-05.c

bool canChange(char* start, char* target) {
    int n = strlen(start);
    int i = 0, j = 0;

    for (int k = 0; k < n; k++) {
        if (start[k] != '_') {
            while (j < n && target[j] == '_') j++;

            if (j == n || start[k] != target[j]) return false;

            if ((start[k] == 'L' && k < j) || (start[k] == 'R' && k > j)) return false;

            j++;
        }
    }

    while (j < n) {
        if (target[j] != '_') return false;
        j++;
    }

    return true;
}

File: December 2024 Leetcode Solution/December-05.rs
impl Solution {
    pub fn can_change(start: String, target: String) -> bool {
        let n = start.len();
        let (mut i, mut j) = (0, 0);
        let start_chars: Vec<char> = start.chars().collect();
        let target_chars: Vec<char> = target.chars().collect();

        while i < n {
            if start_chars[i] != '_' {
                while j < n && target_chars[j] == '_' {
                    j += 1;
                }

                if j == n || start_chars[i] != target_chars[j] {
                    return false;
                }

                if (start_chars[i] == 'L' && i < j) || (start_chars[i] == 'R' && i > j) {
                    return false;
                }

                j += 1;
            }
            i += 1;
        }

        while j < n {
            if target_chars[j] != '_' {
                return false;
            }
            j += 1;
        }

        true
    }
}
